src/pal/src/loader/module.cpp

```diff
@@ -274,6 +276,141 @@ LoadLibraryExW(
     return hModule;
 }
 
+/*
+Function:
+LoadLibraryDirectW
+
+Loads a library using a system call, without registering the library with the module list.
+
+Returns the system handle to the loaded library, or nullptr upon failure (error is set via SetLastError()).
+*/
+HMODULE
+PALAPI
+PAL_LoadLibraryDirect(
+    IN LPCWSTR lpLibFileName)
+{
+    CHAR lpstr[MAX_LONGPATH];
+    INT name_length;
+    HMODULE hModule = NULL;
+
+    PERF_ENTRY(LoadLibraryDirectW);
+    ENTRY("LoadLibraryDirectW (lpLibFileName=%p (%S)) \n",
+          lpLibFileName?lpLibFileName:W16_NULLSTRING,
+          lpLibFileName?lpLibFileName:W16_NULLSTRING);
+
+    if (NULL == lpLibFileName)
+    {
+        ERROR("lpLibFileName is NULL;Exit.\n");
+        SetLastError(ERROR_MOD_NOT_FOUND);
+        goto done;
+    }
+
+    if (lpLibFileName[0] == 0)
+    {
+        ERROR("Can't load library with NULL file name...\n");
+        SetLastError(ERROR_INVALID_PARAMETER);
+        goto done;
+    }
+
+    /* do the Dos/Unix conversion on our own copy of the name */
+
+    name_length = WideCharToMultiByte(CP_ACP, 0, lpLibFileName, -1, lpstr,
+                                      MAX_LONGPATH, NULL, NULL);
+    if (name_length == 0)
+    {
+        DWORD dwLastError = GetLastError();
+        if (dwLastError == ERROR_INSUFFICIENT_BUFFER)
+        {
+            ERROR("lpLibFileName is larger than MAX_LONGPATH (%d)!\n", MAX_LONGPATH);
+        }
+        else
+        {
+            ASSERT("WideCharToMultiByte failure! error is %d\n", dwLastError);
+        }
+        SetLastError(ERROR_INVALID_PARAMETER);
+        goto done;
+    }
+
+    FILEDosToUnixPathA(lpstr);
+
+    /* let LOADLoadLibraryDirect call SetLastError in case of failure */
+    hModule = LOADLoadLibraryDirect(lpstr, true /* setLastError */);
+
+done:
+    LOGEXIT("LoadLibraryDirectW returns HMODULE %p\n", hModule);
+    PERF_EXIT(LoadLibraryDirectW);
+    return hModule;
+}
+
+/*
+Function:
+LoadLibraryDirectW
+
+Loads a library using a system call, without registering the library with the module list.
+
+Returns the system handle to the loaded library, or nullptr upon failure (error is set via SetLastError()).
+*/
+HMODULE
+PALAPI
+PAL_RegisterLibraryDirect(
```

Comment:
This function shares a lot of code with the previous one. Could you please refactor it to use a common helper for the lpLibFileName processing?